bootstrap/handlers/clients.go

```diff
@@ -0,0 +1,149 @@
+// This handler will fail if an unknown client is specified.
+func (cb *ClientsBootstrap) BootstrapHandler(
+	_ context.Context,
+	_ *sync.WaitGroup,
+	startupTimer startup.Timer,
+	dic *di.Container) bool {
+
+	lc := container.LoggingClientFrom(dic.Get)
+	config := container.ConfigurationFrom(dic.Get)
+	cb.registry = container.RegistryFrom(dic.Get)
+
+	for serviceKey, serviceInfo := range config.GetBootstrap().Clients {
+		url, err := cb.getClientUrl(serviceKey, serviceInfo.Url(), startupTimer, lc)
+		if err != nil {
+			lc.Error(err.Error())
+			return false
+		}
+
+		switch serviceKey {
+		case common.CoreDataServiceKey:
+			dic.Update(di.ServiceConstructorMap{
+				container.EventClientName: func(get di.Get) interface{} {
+					return clients.NewEventClient(url)
+				},
+			})
+		case common.CoreMetaDataServiceKey:
+			dic.Update(di.ServiceConstructorMap{
+				container.DeviceClientName: func(get di.Get) interface{} {
+					return clients.NewDeviceClient(url)
+				},
+				container.DeviceServiceClientName: func(get di.Get) interface{} {
+					return clients.NewDeviceServiceClient(url)
+				},
+				container.DeviceProfileClientName: func(get di.Get) interface{} {
+					return clients.NewDeviceProfileClient(url)
+				},
+				container.ProvisionWatcherClientName: func(get di.Get) interface{} {
+					return clients.NewProvisionWatcherClient(url)
+				},
+			})
+
+		case common.CoreCommandServiceKey:
+			dic.Update(di.ServiceConstructorMap{
+				container.CommandClientName: func(get di.Get) interface{} {
+					return clients.NewCommandClient(url)
+				},
+			})
+
+		case common.SupportNotificationsServiceKey:
+			dic.Update(di.ServiceConstructorMap{
+				container.NotificationClientName: func(get di.Get) interface{} {
+					return clients.NewNotificationClient(url)
+				},
+				container.SubscriptionClientName: func(get di.Get) interface{} {
+					return clients.NewSubscriptionClient(url)
+				},
+			})
+
+		case common.SupportSchedulerServiceKey:
+			dic.Update(di.ServiceConstructorMap{
+				container.IntervalClientName: func(get di.Get) interface{} {
+					return clients.NewIntervalClient(url)
+				},
+				container.IntervalActionClientName: func(get di.Get) interface{} {
+					return clients.NewIntervalActionClient(url)
+				},
+			})
+
+		default:
+
+		}
+	}
+
+	return true
+}
+
+func (cb *ClientsBootstrap) getClientUrl(serviceKey string, defaultUrl string, startupTimer startup.Timer, lc logger.LoggingClient) (string, error) {
+	if cb.registry == nil {
+		lc.Debugf("Using configuration for URL for '%s': %s", serviceKey, defaultUrl)
+		return defaultUrl, nil
+	}
+
+	var err error
+	var endpoint types.ServiceEndpoint
+
+	for startupTimer.HasNotElapsed() {
+		endpoint, err = cb.registry.GetServiceEndpoint(serviceKey)
+		if err == nil {
+			break
+		}
+
+		startupTimer.SleepForInterval()
+	}
+
+	if err != nil {
+		return "", fmt.Errorf("unable to Get service endpoint for '%s': %s", serviceKey, err.Error())
```

Comment:
The error  is returned to the caller which will log it as an error.
In retrospect, those log message should be info since this only occurs once on start-up. Will make that change.
The startupTimer doesn't have an API to get the time waited, which is default to 60secs and and can be overridden by env var.
https://docs.edgexfoundry.org/2.1/microservices/configuration/CommonEnvironmentVariables/#edgex_startup_duration